# The main entry point of your workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.

#workdir: "/cluster/home/quever/workflows/initialize/metagenomics/"
workdir: "/cluster/projects/mcgahalab/data/mcgahalab/PASS/human"

include: "rules/common.smk"
include: "rules/other.smk"
include: "rules/kneaddata.smk"
include: "rules/metaphlan.smk"
include: "rules/humann.smk"
include: "rules/strainphlan.smk"

rule all:
    input:
        expand("results/kneaddata/main/{sample}.knead1.fastq.gz", sample = samples["sample_name"]),
        expand("results/metaphlan/main/{sample}.s1.tsv", sample = samples["sample_name"]),
        "results/metaphlan/merged/merged_metaphlan.count.tsv"
#        "results/metaphlan/merged/merged_metaphlan.count.tsv",
#        "results/metaphlan/merged/merged_metaphlan.relab.tsv",
#        "results/humann/merged/ecs.tsv",
#        "results/humann/merged/genefamilies.tsv",
#        "results/humann/merged/pathabundance.tsv",
#        "results/humann/merged/ecs.relab.tsv",
#        "results/humann/merged/genefamilies.relab.tsv",
#        "results/humann/merged/pathabundance.relab.tsv",
#        "results/strainphlan/consensus_markers/clades_list_order_by_average_abundance.txt",
#        "results/strainphlan/strainphlan.out",
